walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,112 @@
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.Enumeration;
+import java.util.Map;
+import java.util.Iterator;
+
+import java.util.Comparator;
 
 public final class FindMeetingQuery {
+  /** Return free slots based on list of time points. */
+  public Collection<TimeRange> getFreeSlots(List<TimePoint> importantEvents, MeetingRequest request) {
+    Collection<TimeRange> possibleSlots = new ArrayList<>();
+    // Sort all time points.
+    Collections.sort(importantEvents, TimePoint.ORDER_END_START);
+
+    // Number of active meeting for the moment under consideration.
+    int numberOfActiveMeetings = 0;
+    int endOfLastMeeting = 0;
+
+    // numberOfActiveMeetings will be 0  - before start and after end of the day,
+    // 1 - if current time slot is free for all mandatory attendees,
+    // >1 - in any other case.
+  
+    // numberOfActiveMeetings will increase if new time point under consideration is a start for a meeting, and decrease in other case.
+    for(TimePoint event : importantEvents) {
+      if (numberOfActiveMeetings == 1) {
+        // Check if there is enough time for requested meeting.
+        if (event.time() - endOfLastMeeting >= request.getDuration()) {
+          TimeRange freeSlot = TimeRange.fromStartDuration(endOfLastMeeting, event.time() - endOfLastMeeting);
+          possibleSlots.add(freeSlot);
+        }
+      }
+      
+      if (event.type() == 1) {
+        numberOfActiveMeetings += 1;
+      } else {
+        numberOfActiveMeetings -= 1;
+        endOfLastMeeting = event.time();
+      }
+    }
+    return possibleSlots;
+  }
+
+  /** Return list of time slots for current request for mandatory and if possible for optional attendees. */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Create a list of existing meetings with mandatory attendees.
+    List<TimePoint> importantEvents = new ArrayList<>();
+    // Create a list of existing meetings with optional attendees considered as mandatory.
+    List<TimePoint> optionalEvents = new ArrayList<>();
+  
+    // List of possible time slots for mandatory attendees.
+    Collection<TimeRange> possibleMandatorySlots = new ArrayList<>();
+    // List of possible time slots for optional attendees.
+    Collection<TimeRange> possibleOptionalSlots = new ArrayList<>();
+
+    // Add start and end of the day as time points.
+    TimePoint startOfTheDay = new TimePoint(0, 1, -1);
+    TimePoint endOfTheDay = new TimePoint(24 * 60, 2, -1);
+
+    importantEvents.add(startOfTheDay);
+    importantEvents.add(endOfTheDay);
+
+    optionalEvents.add(startOfTheDay);
+    optionalEvents.add(endOfTheDay);
+
+    for (Event event : events) {
+      boolean mandatoryAttendee = false;
+      // Pick events with mandatory attendees (fill list for optional as well).
+      for (String attendee: request.getAttendees()) {
```

Comment:
optional: You can probably replace this loop with something like https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#disjoint-java.util.Collection-java.util.Collection- or https://docs.oracle.com/javase/8/docs/api/java/util/Set.html#retainAll-java.util.Collection-

Reply:
Replaced both loops with disjoint function.